File: BOJ_SOL/1068.cpp
#include <iostream>
#include <vector>

using namespace std;

int n, del_node;
vector<int> nodes, graph[50];

void delete_node(int node) {
    for(int i = 0; i < graph[node].size(); i++){
        int child = graph[node][i];
        delete_node(child);
    }
    graph[node].clear();
    for (int i = 0; i < n; i++){
        for (int j = 0; j < graph[i].size(); j++){
            if(graph[i][j] == node){
                graph[i].erase(graph[i].begin() + j);
                break;
            }
        }
    }
}

int find_root(const vector<int> &nodes){
    for (int i = 0; i < nodes.size(); i++){
        if(nodes[i] == -1){
            return i;
        }
    }
    return -1;
}

int count_leaf(const vector<int> graph[], int v){
    if(graph[v].empty()){
        return 1;
    }
    int cnt = 0;
    for(int i = 0; i < graph[v].size(); i++){
        cnt += count_leaf(graph, graph[v][i]);
    }
    return cnt;
}

int main(){
    cin >> n;
    nodes.resize(n);
    for (int i = 0; i < n; i++){
        cin >> nodes[i];
        if(nodes[i] != -1){
            graph[nodes[i]].push_back(i);
        }
    }
    cin >> del_node;
    delete_node(del_node);
    int root = find_root(nodes);
    if(del_node == root){
        cout << 0 << endl;
    }
    else{
        cout << count_leaf(graph, root) << endl;
    }
    return 0;
}

File: BOJ_SOL/1068.py
import sys
input = sys.stdin.readline
#입력값 받기와 기본 그래프 생성
n = int(input())
nodes = list(map(int, input().split()))
del_node = int(input())
graph = [[] for _ in range(n)]
#노드의 부모 정보들로 인접리스트 만들기
for i in range(n):
    if nodes[i] != -1:
        graph[nodes[i]].append(i)
#그래프에서 노드 제거하기
def delete(node):
    for i in graph[node]:
        delete(i)
    graph[node] = []
    for i in range(n):
        if node in graph[i]:
            graph[i].remove(node)

delete(del_node)

# 루트 노드를 찾는 함수
def find_root(nodes):
    return nodes.index(-1)

# 루트 노드부터 탐색하여 리프 노드의 개수를 계산
def count_leaf(graph, v):
    if len(graph[v]) == 0:
        return 1
    cnt = 0
    for i in graph[v]:
        cnt += count_leaf(graph, i)
    return cnt

#루트를 찾고, 루트를 없애는 문제라면 바로 0을 반환하게 함
root = find_root(nodes)
print(count_leaf(graph, root) if del_node != root else 0)
